### PR TITLE
Add support for Core-Config

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
Operand NFD Patch https://github.com/kubernetes-sigs/node-feature-discovery/pull/433  changed the way nfd-worker get's arguments, this patch looks to update the operator to the new way of running the nfd-worker 

Signed-off-by: Carlos Eduardo Arango Gutierrez <carangog@redhat.com>

### CODE REVIEW

File: config/manager/manager.yaml [hunk 1]
```diff
@@ -30,7 +30,7 @@ spec:
             capabilities:
               drop: ["ALL"]
           command:
-            - /manager
+            - /node-feature-discovery-operator
```
Comment (marquiz) [inline config/manager/manager.yaml:33]: nit: this is another unrelated change

File: config/samples/nfd.kubernetes.io_v1_nodefeaturediscovery.yaml [hunk 2]
```diff
@@ -6,10 +6,30 @@ metadata:
 spec:
   operand:
     namespace: node-feature-discovery-operator
-    image: quay.io/openshift/origin-node-feature-discovery:4.7
+    image: gcr.io/k8s-staging-nfd/node-feature-discovery:master
```
Comment (marquiz) [inline config/samples/nfd.kubernetes.io_v1_nodefeaturediscovery.yaml:9]: What image?

File: controllers/nodefeaturediscovery_state.go [hunk 3]
```diff
@@ -44,8 +44,10 @@ func (n *NFD) init(
 	n.rec = r
 	n.ins = i
 	n.idx = 0
-	n.addState("/opt/nfd/master")
-	n.addState("/opt/nfd/worker")
+	if len(n.controls) == 0 {
+		n.addState("/opt/nfd/master")
+		n.addState("/opt/nfd/worker")
+	}
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_state.go:50]: unrelated?

File: config/crd/bases/nfd.kubernetes.io_nodefeaturediscoveries.yaml [hunk 4]
```diff
@@ -38,6 +38,16 @@ spec:
           spec:
             description: NodeFeatureDiscoverySpec defines the desired state of NodeFeatureDiscovery
             properties:
+              customConfig:
+                description: ConfigMap describes configuration options for the NFD
```
Comment (marquiz) [inline config/crd/bases/nfd.kubernetes.io_nodefeaturediscoveries.yaml:42]: Not true, this will only contain rules for the custom source

**Review by marquiz — CHANGES_REQUESTED** [review]
@ArangoGutierrez thanks for the PR!

Generally, looks good - could you just drop the unrelated pieces and address those in separate PRs? 🥺 I think that in general one PR (or at least patch) should "do one thing" - this will make the reviewer's task a lot easier and also contribute to a more comprehensible commit history.

I also started to ponder the custom config. Those are designed for extra custom config fragments (e.g. from 3rd party) and the idea is that there can be multiple of those. I think we should either drop those at this point (custom rules can be injected directly in the main config) or make the custom configs an array (i.e. support multiple ones). WDYT?

File: config/manager/manager.yaml (hunk 1 above)
Comment (ArangoGutierrez) [reply config/manager/manager.yaml:33]: https://github.com/kubernetes-sigs/node-feature-discovery-operator/pull/57

File: controllers/nodefeaturediscovery_state.go (hunk 3 above)
Comment (ArangoGutierrez) [reply controllers/nodefeaturediscovery_state.go:50]: https://github.com/kubernetes-sigs/node-feature-discovery-operator/pull/57

File: Makefile [hunk 5]
```diff
@@ -1,10 +1,4 @@
-IMAGE_BUILD_CMD ?= docker build
-IMAGE_BUILD_EXTRA_OPTS ?=
-IMAGE_PUSH_CMD ?= docker push
-CONTAINER_RUN_CMD ?= docker run -u "`id -u`:`id -g`"
```
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply Makefile:4]: sure, going to leave `CONTAINER_RUN_CMD ?= docker run -u "`id -u`:`id -g`"` the others are repeated on line 48, that is only why I am cleaning the makefile. 
If you prefer I can add this Makefile cleanup to the #53 unrelated minor fixes PR ?

File: config/crd/bases/nfd.kubernetes.io_nodefeaturediscoveries.yaml (hunk 4 above)
Comment (ArangoGutierrez) [reply config/crd/bases/nfd.kubernetes.io_nodefeaturediscoveries.yaml]: fixed

File: config/samples/nfd.kubernetes.io_v1_nodefeaturediscovery.yaml (hunk 2 above)
Comment (ArangoGutierrez) [reply config/samples/nfd.kubernetes.io_v1_nodefeaturediscovery.yaml]: fixed

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply Makefile:4]: fixed

**Comment by ArangoGutierrez** [comment]
> an be multiple of those. I think we should either drop those at this point (custom rules can be injected directly in the main config) or make the custom configs an array (i.e. support multiple ones). WDYT?

I think that will require discussion and an extra (following) PR, the idea with operators is to always handle things via CRD's, prevent users going and creating their own config maps and injecting them into pods.  I can see how we could create more mini config maps and hold an array. but I would like to see this PR through and address the multiple custom rules in a separate PR
WDYT?

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (marquiz) [reply Makefile:4]: Could you submit the changes to the Makefile as a separate PR? We could/should review/merge it separately

**Comment by marquiz** [comment]
> I think that will require discussion and an extra (following) PR, the idea with operators is to always handle things via CRD's, prevent users going and creating their own config maps and injecting them into pods. I can see how we could create more mini config maps and hold an array. but I would like to see this PR through and address the multiple custom rules in a separate PR
> WDYT?

I don't necessarily see the usefulness of of a single `custom-config` in the operator. Might be just a source of unnecessary confusion as the user can equally well inject the custom rules (that would go into custom-config) into the main worker config. With an array (multiple custom configs) I can see a use case of dropping custom config fragments from various sources independently, without altering any other configs. Thus, I'd suggest to drop it this for now and address it in a separate PR. How does that sound? Any thoughts from @zvonkok ?

File: config/samples/nfd.kubernetes.io_v1_nodefeaturediscovery.yaml [hunk 6]
```diff
@@ -6,7 +6,7 @@ metadata:
 spec:
   operand:
     namespace: node-feature-discovery-operator
-    image: gcr.io/k8s-staging-nfd/node-feature-discovery:master
+    image: gcr.io/k8s-staging-nfd/node-feature-discovery:v0.8.1
```
[the remark this answers is not in the document]
Comment (marquiz) [reply config/samples/nfd.kubernetes.io_v1_nodefeaturediscovery.yaml:9]: Actually, I'd submit this change in a separate PR, too.

...and use
```
image: gcr.io/k8s-staging-nfd/node-feature-discovery:master
```

[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply config/samples/nfd.kubernetes.io_v1_nodefeaturediscovery.yaml:9]: https://github.com/kubernetes-sigs/node-feature-discovery-operator/pull/58

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply Makefile:4]: https://github.com/kubernetes-sigs/node-feature-discovery-operator/pull/59

**Comment by ArangoGutierrez** [comment]
> > I think that will require discussion and an extra (following) PR, the idea with operators is to always handle things via CRD's, prevent users going and creating their own config maps and injecting them into pods. I can see how we could create more mini config maps and hold an array. but I would like to see this PR through and address the multiple custom rules in a separate PR
> > WDYT?
> 
> I don't necessarily see the usefulness of of a single `custom-config` in the operator. Might be just a source of unnecessary confusion as the user can equally well inject the custom rules (that would go into custom-config) into the main worker config. With an array (multiple custom configs) I can see a use case of dropping custom config fragments from various sources independently, without altering any other configs. Thus, I'd suggest to drop it this for now and address it in a separate PR. How does that sound? Any thoughts from @zvonkok ?

Drop `custom-config` from this PR, leave this PR to be just for `core-config` and address/discuss `custom-config` on it's won PR? makes sense to me

**Comment by ArangoGutierrez** [comment]
needs rebase

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply Makefile:4]: rebased
